src/Voron/Impl/Journal/WriteAheadJournal.cs

```diff
@@ -234,12 +234,32 @@ public bool RecoverDatabase(TransactionHeader* txHeader, Action<string> addToIni
                 // corruption when applying journals at recovery time rather than at usage.
                 var tempTx = new TempPagerTransaction();
 
-                foreach (var modifiedPage in modifiedPages)
+                var sortedPages = modifiedPages.ToArray();
+                Array.Sort(sortedPages);
+
+                long minPageChecked = -1;
+
+                // we need to iterate from the end in order to filter out pages that was overwritten by later transaction
+
+                for (var i = sortedPages.Length - 1; i >= 0; i--)
                 {
-                    var ptr = (PageHeader*)_dataPager.AcquirePagePointerWithOverflowHandling(tempTx, modifiedPage, null);
-                    _env.ValidatePageChecksum(modifiedPage, ptr);
+                    using (tempTx) // release any resources, we just wanted to validate things
+                    {
+                        var modifiedPage = sortedPages[i];
+
+                        var ptr = (PageHeader*)_dataPager.AcquirePagePointerWithOverflowHandling(tempTx, modifiedPage, null);
+
+                        var maxPageRange = modifiedPage + VirtualPagerLegacyExtensions.GetNumberOfPages(ptr) - 1;
 
-                    tempTx.Dispose(); // release any resources, we just wanted to validate things
+                        if (minPageChecked != -1 && maxPageRange >= minPageChecked)
+                        {
+                            continue;
```

Comment:
I wonder if we should do better and ensure that `modifiedPage` was really modified in earlier transaction (by collecting not only the page numbers by also ids of transactions when a page was modified). 

@ayende ?

Reply:
We don't care about that, just that the value of the page is valid.